File: microsoft_auth_helper.py
"""
Microsoft Authentication Helper
Handles Microsoft account authentication with proper error handling for sub status 6008
"""

import json
import logging
from datetime import datetime
from typing import Dict, Optional, List
from enum import Enum


class AuthErrorCode(Enum):
    """Microsoft authentication error codes"""
    INVALID_CREDENTIAL = 6008
    ACCOUNT_LOCKED = 6009
    PASSWORD_EXPIRED = 6010
    MFA_REQUIRED = 50076
    CONDITIONAL_ACCESS_BLOCKED = 53003


class MicrosoftAuthError(Exception):
    """Custom exception for Microsoft authentication errors"""

    def __init__(
        self,
        message: str,
        sub_status: int,
        correlation_id: str,
        error_code: Optional[int] = None,
        timestamp: Optional[str] = None
    ):
        self.message = message
        self.sub_status = sub_status
        self.correlation_id = correlation_id
        self.error_code = error_code or 2150171662
        self.timestamp = timestamp or datetime.utcnow().isoformat()
        super().__init__(self.message)

    def __str__(self):
        return (
            f"MicrosoftAuthError(code={self.error_code}, "
            f"sub_status={self.sub_status}, "
            f"correlation_id={self.correlation_id}): {self.message}"
        )


class MicrosoftAuthHelper:
    """Helper class for Microsoft authentication with comprehensive error handling"""

    def __init__(self, log_level=logging.INFO):
        self.logger = logging.getLogger(__name__)
        self.logger.setLevel(log_level)

        if not self.logger.handlers:
            handler = logging.StreamHandler()
            formatter = logging.Formatter(
                '%(asctime)s - %(name)s - %(levelname)s - %(message)s'
            )
            handler.setFormatter(formatter)
            self.logger.addHandler(handler)

    def handle_auth_error(
        self,
        error: Exception,
        correlation_id: Optional[str] = None
    ) -> Dict:
        """
        Handle Microsoft authentication errors with specific handling for sub status 6008

        Args:
            error: The exception raised during authentication
            correlation_id: Optional correlation ID for tracking

        Returns:
            Dict containing error details and suggested actions
        """
        # Extract error details
        sub_status = getattr(error, 'sub_status', None)
        error_code = getattr(error, 'error_code', None)
        correlation_id = correlation_id or getattr(error, 'correlation_id', 'N/A')

        self.logger.error(
            f"Authentication error - Code: {error_code}, "
            f"Sub Status: {sub_status}, "
            f"Correlation ID: {correlation_id}"
        )

        # Handle sub status 6008 specifically
        if sub_status == 6008 or sub_status == AuthErrorCode.INVALID_CREDENTIAL.value:
            return self._handle_invalid_credential_error(correlation_id, error)

        # Handle other common errors
        error_handlers = {
            6009: self._handle_account_locked_error,
            6010: self._handle_password_expired_error,
            50076: self._handle_mfa_required_error,
            53003: self._handle_conditional_access_blocked_error
        }

        handler = error_handlers.get(sub_status)
        if handler:
            return handler(correlation_id, error)

        # Generic error handling
        return self._handle_generic_error(correlation_id, error)

    def _handle_invalid_credential_error(
        self,
        correlation_id: str,
        error: Exception
    ) -> Dict:
        """Handle invalid credential error (sub status 6008)"""
        self.logger.error("Invalid credential error detected (Sub Status 6008)")

        suggestions = [
            "Verify that your username and password are correct",
            "Check if your account has been locked or disabled",
            "Clear cached credentials and try again",
            "Reset your password at https://account.live.com/password/reset",
            "Verify your account at https://account.microsoft.com",
            "Check if MFA (Multi-Factor Authentication) is configured correctly",
            "If in an organization, contact your IT administrator",
            "Try using a different browser or incognito mode",
            "Ensure your system time is synchronized correctly"
        ]

        troubleshooting_steps = [
            {
                "step": 1,
                "action": "Reset Password",
                "description": "Visit https://account.live.com/password/reset"
            },
            {
                "step": 2,
                "action": "Check Account Status",
                "description": "Verify account is not locked at https://account.microsoft.com"
            },
            {
                "step": 3,
                "action": "Clear Credentials",
                "description": "Clear cached credentials and browser cookies"
            },
            {
                "step": 4,
                "action": "Verify MFA",
                "description": "Check MFA settings at https://account.microsoft.com/security"
            }
        ]

        return {
            'status': 'error',
            'error_type': 'INVALID_CREDENTIAL',
            'error_code': 2150171662,
            'sub_status': 6008,
            'message': 'The credential is invalid. Unexpected sub status (6008)',
            'correlation_id': correlation_id,
            'timestamp': datetime.utcnow().isoformat(),
            'suggestions': suggestions,
            'troubleshooting_steps': troubleshooting_steps,
            'support_info': {
                'url': 'https://support.microsoft.com',
                'note': 'Provide correlation ID when contacting support'
            }
        }

    def _handle_account_locked_error(
        self,
        correlation_id: str,
        error: Exception
    ) -> Dict:
        """Handle account locked error"""
        return {
            'status': 'error',
            'error_type': 'ACCOUNT_LOCKED',
            'sub_status': 6009,
            'message': 'Account is locked',
            'correlation_id': correlation_id,
            'timestamp': datetime.utcnow().isoformat(),
            'suggestions': [
                'Wait 30 minutes and try again',
                'Contact your administrator',
                'Visit https://account.microsoft.com to unlock'
            ]
        }

    def _handle_password_expired_error(
        self,
        correlation_id: str,
        error: Exception
    ) -> Dict:
        """Handle password expired error"""
        return {
            'status': 'error',
            'error_type': 'PASSWORD_EXPIRED',
            'sub_status': 6010,
            'message': 'Password has expired',
            'correlation_id': correlation_id,
            'timestamp': datetime.utcnow().isoformat(),
            'suggestions': [
                'Reset your password',
                'Visit https://account.live.com/password/reset'
            ]
        }

    def _handle_mfa_required_error(
        self,
        correlation_id: str,
        error: Exception
    ) -> Dict:
        """Handle MFA required error"""
        return {
            'status': 'error',
            'error_type': 'MFA_REQUIRED',
            'sub_status': 50076,
            'message': 'Multi-factor authentication is required',
            'correlation_id': correlation_id,
            'timestamp': datetime.utcnow().isoformat(),
            'suggestions': [
                'Complete MFA challenge',
                'Check your authenticator app or phone',
                'Ensure you have access to your MFA method'
            ]
        }

    def _handle_conditional_access_blocked_error(
        self,
        correlation_id: str,
        error: Exception
    ) -> Dict:
        """Handle conditional access blocked error"""
        return {
            'status': 'error',
            'error_type': 'CONDITIONAL_ACCESS_BLOCKED',
            'sub_status': 53003,
            'message': 'Access blocked by conditional access policy',
            'correlation_id': correlation_id,
            'timestamp': datetime.utcnow().isoformat(),
            'suggestions': [
                'Contact your IT administrator',
                'Verify your device is compliant',
                'Check if you are using an approved network/location',
                'Ensure your browser/application is approved'
            ]
        }

    def _handle_generic_error(
        self,
        correlation_id: str,
        error: Exception
    ) -> Dict:
        """Handle generic authentication error"""
        return {
            'status': 'error',
            'error_type': 'GENERIC_AUTH_ERROR',
            'message': str(error),
            'correlation_id': correlation_id,
            'timestamp': datetime.utcnow().isoformat(),
            'suggestions': [
                'Check your internet connection',
                'Try again later',
                'Contact support with correlation ID'
            ]
        }

    def log_error_to_file(self, error_details: Dict, filename: str = 'auth_errors.log'):
        """Log error details to a file for debugging"""
        try:
            with open(filename, 'a') as f:
                f.write(json.dumps(error_details, indent=2))
                f.write('\n---\n')
            self.logger.info(f"Error logged to {filename}")
        except Exception as e:
            self.logger.error(f"Failed to log error to file: {e}")


# Example usage
if __name__ == "__main__":
    # Simulate the error from the user's report
    auth_helper = MicrosoftAuthHelper()

    # Create a simulated error with sub status 6008
    simulated_error = MicrosoftAuthError(
        message="The credential is invalid. Unexpected sub status (6008).",
        sub_status=6008,
        correlation_id="10672792-8bca-4953-b9de-297cd77572c5",
        error_code=2150171662,
        timestamp="2025-12-18T02:41:06.000Z"
    )

    # Handle the error
    error_response = auth_helper.handle_auth_error(
        simulated_error,
        correlation_id="10672792-8bca-4953-b9de-297cd77572c5"
    )

    # Print the detailed error response
    print("\n" + "="*80)
    print("Microsoft Authentication Error Analysis")
    print("="*80)
    print(json.dumps(error_response, indent=2, ensure_ascii=False))
    print("="*80)

    # Log to file
    auth_helper.log_error_to_file(error_response)

    print("\nError details have been logged to 'auth_errors.log'")
    print(f"\nCorrelation ID for support: {error_response['correlation_id']}")

File: microsoft-auth-helper.ts
/**
 * Microsoft Authentication Helper
 * Handles Microsoft account authentication with proper error handling for sub status 6008
 */

export enum AuthErrorCode {
  INVALID_CREDENTIAL = 6008,
  ACCOUNT_LOCKED = 6009,
  PASSWORD_EXPIRED = 6010,
  MFA_REQUIRED = 50076,
  CONDITIONAL_ACCESS_BLOCKED = 53003,
}

export interface AuthError {
  message: string;
  subStatus: number;
  correlationId: string;
  errorCode?: number;
  timestamp?: string;
}

export interface ErrorResponse {
  status: string;
  errorType: string;
  errorCode?: number;
  subStatus: number;
  message: string;
  correlationId: string;
  timestamp: string;
  suggestions: string[];
  troubleshootingSteps?: TroubleshootingStep[];
  supportInfo?: SupportInfo;
}

export interface TroubleshootingStep {
  step: number;
  action: string;
  description: string;
}

export interface SupportInfo {
  url: string;
  note: string;
}

export class MicrosoftAuthError extends Error {
  public subStatus: number;
  public correlationId: string;
  public errorCode: number;
  public timestamp: string;

  constructor(
    message: string,
    subStatus: number,
    correlationId: string,
    errorCode: number = 2150171662,
    timestamp: string = new Date().toISOString()
  ) {
    super(message);
    this.name = 'MicrosoftAuthError';
    this.subStatus = subStatus;
    this.correlationId = correlationId;
    this.errorCode = errorCode;
    this.timestamp = timestamp;
  }

  toString(): string {
    return `${this.name}(code=${this.errorCode}, sub_status=${this.subStatus}, correlation_id=${this.correlationId}): ${this.message}`;
  }
}

export class MicrosoftAuthHelper {
  private logLevel: string;

  constructor(logLevel: string = 'info') {
    this.logLevel = logLevel;
  }

  /**
   * Handle Microsoft authentication errors with specific handling for sub status 6008
   */
  public handleAuthError(
    error: Error | MicrosoftAuthError,
    correlationId?: string
  ): ErrorResponse {
    // Extract error details
    const authError = error as MicrosoftAuthError;
    const subStatus = authError.subStatus || 0;
    const errorCode = authError.errorCode || 0;
    const corrId = correlationId || authError.correlationId || 'N/A';

    console.error(
      `Authentication error - Code: ${errorCode}, ` +
      `Sub Status: ${subStatus}, ` +
      `Correlation ID: ${corrId}`
    );

    // Handle sub status 6008 specifically
    if (subStatus === 6008 || subStatus === AuthErrorCode.INVALID_CREDENTIAL) {
      return this.handleInvalidCredentialError(corrId, error);
    }

    // Handle other common errors
    const errorHandlers: { [key: number]: (id: string, err: Error) => ErrorResponse } = {
      [AuthErrorCode.ACCOUNT_LOCKED]: this.handleAccountLockedError.bind(this),
      [AuthErrorCode.PASSWORD_EXPIRED]: this.handlePasswordExpiredError.bind(this),
      [AuthErrorCode.MFA_REQUIRED]: this.handleMfaRequiredError.bind(this),
      [AuthErrorCode.CONDITIONAL_ACCESS_BLOCKED]: this.handleConditionalAccessBlockedError.bind(this),
    };

    const handler = errorHandlers[subStatus];
    if (handler) {
      return handler(corrId, error);
    }

    // Generic error handling
    return this.handleGenericError(corrId, error);
  }

  /**
   * Handle invalid credential error (sub status 6008)
   */
  private handleInvalidCredentialError(
    correlationId: string,
    error: Error
  ): ErrorResponse {
    console.error('Invalid credential error detected (Sub Status 6008)');

    const suggestions = [
      'Verify that your username and password are correct',
      'Check if your account has been locked or disabled',
      'Clear cached credentials and try again',
      'Reset your password at https://account.live.com/password/reset',
      'Verify your account at https://account.microsoft.com',
      'Check if MFA (Multi-Factor Authentication) is configured correctly',
      'If in an organization, contact your IT administrator',
      'Try using a different browser or incognito mode',
      'Ensure your system time is synchronized correctly',
    ];

    const troubleshootingSteps: TroubleshootingStep[] = [
      {
        step: 1,
        action: 'Reset Password',
        description: 'Visit https://account.live.com/password/reset',
      },
      {
        step: 2,
        action: 'Check Account Status',
        description: 'Verify account is not locked at https://account.microsoft.com',
      },
      {
        step: 3,
        action: 'Clear Credentials',
        description: 'Clear cached credentials and browser cookies',
      },
      {
        step: 4,
        action: 'Verify MFA',
        description: 'Check MFA settings at https://account.microsoft.com/security',
      },
    ];

    return {
      status: 'error',
      errorType: 'INVALID_CREDENTIAL',
      errorCode: 2150171662,
      subStatus: 6008,
      message: 'The credential is invalid. Unexpected sub status (6008)',
      correlationId,
      timestamp: new Date().toISOString(),
      suggestions,
      troubleshootingSteps,
      supportInfo: {
        url: 'https://support.microsoft.com',
        note: 'Provide correlation ID when contacting support',
      },
    };
  }

  /**
   * Handle account locked error
   */
  private handleAccountLockedError(
    correlationId: string,
    error: Error
  ): ErrorResponse {
    return {
      status: 'error',
      errorType: 'ACCOUNT_LOCKED',
      subStatus: 6009,
      message: 'Account is locked',
      correlationId,
      timestamp: new Date().toISOString(),
      suggestions: [
        'Wait 30 minutes and try again',
        'Contact your administrator',
        'Visit https://account.microsoft.com to unlock',
      ],
    };
  }

  /**
   * Handle password expired error
   */
  private handlePasswordExpiredError(
    correlationId: string,
    error: Error
  ): ErrorResponse {
    return {
      status: 'error',
      errorType: 'PASSWORD_EXPIRED',
      subStatus: 6010,
      message: 'Password has expired',
      correlationId,
      timestamp: new Date().toISOString(),
      suggestions: [
        'Reset your password',
        'Visit https://account.live.com/password/reset',
      ],
    };
  }

  /**
   * Handle MFA required error
   */
  private handleMfaRequiredError(
    correlationId: string,
    error: Error
  ): ErrorResponse {
    return {
      status: 'error',
      errorType: 'MFA_REQUIRED',
      subStatus: 50076,
      message: 'Multi-factor authentication is required',
      correlationId,
      timestamp: new Date().toISOString(),
      suggestions: [
        'Complete MFA challenge',
        'Check your authenticator app or phone',
        'Ensure you have access to your MFA method',
      ],
    };
  }

  /**
   * Handle conditional access blocked error
   */
  private handleConditionalAccessBlockedError(
    correlationId: string,
    error: Error
  ): ErrorResponse {
    return {
      status: 'error',
      errorType: 'CONDITIONAL_ACCESS_BLOCKED',
      subStatus: 53003,
      message: 'Access blocked by conditional access policy',
      correlationId,
      timestamp: new Date().toISOString(),
      suggestions: [
        'Contact your IT administrator',
        'Verify your device is compliant',
        'Check if you are using an approved network/location',
        'Ensure your browser/application is approved',
      ],
    };
  }

  /**
   * Handle generic authentication error
   */
  private handleGenericError(
    correlationId: string,
    error: Error
  ): ErrorResponse {
    return {
      status: 'error',
      errorType: 'GENERIC_AUTH_ERROR',
      subStatus: 0,
      message: error.message,
      correlationId,
      timestamp: new Date().toISOString(),
      suggestions: [
        'Check your internet connection',
        'Try again later',
        'Contact support with correlation ID',
      ],
    };
  }

  /**
   * Log error details to console in a formatted way
   */
  public logErrorDetails(errorResponse: ErrorResponse): void {
    console.log('\n' + '='.repeat(80));
    console.log('Microsoft Authentication Error Analysis');
    console.log('='.repeat(80));
    console.log(JSON.stringify(errorResponse, null, 2));
    console.log('='.repeat(80));
    console.log(`\nCorrelation ID for support: ${errorResponse.correlationId}`);
  }

  /**
   * Save error details to local storage (for browser environments)
   */
  public saveErrorToLocalStorage(errorResponse: ErrorResponse): void {
    if (typeof window !== 'undefined' && window.localStorage) {
      const errors = JSON.parse(localStorage.getItem('auth_errors') || '[]');
      errors.push(errorResponse);
      localStorage.setItem('auth_errors', JSON.stringify(errors));
      console.log('Error saved to localStorage');
    }
  }

  /**
   * Retry authentication with exponential backoff
   */
  public async retryWithBackoff<T>(
    fn: () => Promise<T>,
    maxRetries: number = 3,
    initialDelay: number = 1000
  ): Promise<T> {
    let lastError: Error | undefined;

    for (let attempt = 0; attempt < maxRetries; attempt++) {
      try {
        return await fn();
      } catch (error) {
        lastError = error as Error;

        if (attempt < maxRetries - 1) {
          const delay = initialDelay * Math.pow(2, attempt);
          console.log(`Retry attempt ${attempt + 1}/${maxRetries} after ${delay}ms`);
          await this.sleep(delay);
        }
      }
    }

    throw lastError || new Error('Max retries exceeded');
  }

  private sleep(ms: number): Promise<void> {
    return new Promise(resolve => setTimeout(resolve, ms));
  }
}

// Example usage
if (typeof require !== 'undefined' && require.main === module) {
  // Simulate the error from the user's report
  const authHelper = new MicrosoftAuthHelper();

  // Create a simulated error with sub status 6008
  const simulatedError = new MicrosoftAuthError(
    'The credential is invalid. Unexpected sub status (6008).',
    6008,
    '10672792-8bca-4953-b9de-297cd77572c5',
    2150171662,
    '2025-12-18T02:41:06.000Z'
  );

  // Handle the error
  const errorResponse = authHelper.handleAuthError(
    simulatedError,
    '10672792-8bca-4953-b9de-297cd77572c5'
  );

  // Log the detailed error response
  authHelper.logErrorDetails(errorResponse);
}

export default MicrosoftAuthHelper;
